video/out/vo_dmabuf_wayland.c

```diff
@@ -73,34 +73,33 @@ static bool vaapi_dmabuf_importer(struct mp_image *src, struct wlbuf_pool_entry*
                                   struct zwp_linux_buffer_params_v1 *params)
 {
     struct priv *p = entry->vo->priv;
-    VADRMPRIMESurfaceDescriptor desc;
+    VADRMPRIMESurfaceDescriptor desc = { 0 };
     bool dmabuf_imported = false;
     /* composed has single layer */
     int layer_no = 0;
     VAStatus status = vaExportSurfaceHandle(p->display, entry->key, VA_SURFACE_ATTRIB_MEM_TYPE_DRM_PRIME_2,
                                             VA_EXPORT_SURFACE_COMPOSED_LAYERS | VA_EXPORT_SURFACE_READ_ONLY, &desc);
-    if (status == VA_STATUS_ERROR_INVALID_SURFACE) {
-        MP_VERBOSE(entry->vo, "VA export to composed layers not supported.\n");
-    } else if (!vo_wayland_supported_format(entry->vo, desc.layers[0].drm_format, desc.objects[0].drm_format_modifier)) {
-        MP_VERBOSE(entry->vo, "%s(%016lx) is not supported.\n",
-                   mp_tag_str(desc.layers[0].drm_format), desc.objects[0].drm_format_modifier);
-    } else if (CHECK_VA_STATUS(entry->vo, "vaExportSurfaceHandle()")) {
-        entry->drm_format = desc.layers[layer_no].drm_format;
-        for (int plane_no = 0; plane_no < desc.layers[layer_no].num_planes; ++plane_no) {
-            int object = desc.layers[layer_no].object_index[plane_no];
-            uint64_t modifier = desc.objects[object].drm_format_modifier;
-            zwp_linux_buffer_params_v1_add(params, desc.objects[object].fd, plane_no, desc.layers[layer_no].offset[plane_no],
-                                           desc.layers[layer_no].pitch[plane_no], modifier >> 32, modifier & 0xffffffff);
-        }
-        dmabuf_imported = true;
-    }
 
-    /* clean up descriptor */
-    if (status != VA_STATUS_ERROR_INVALID_SURFACE) {
-        for (int i = 0; i < desc.num_objects; i++) {
-            close(desc.objects[i].fd);
-            desc.objects[i].fd = 0;
+    if (status == VA_STATUS_ERROR_INVALID_SURFACE)
+        MP_VERBOSE(entry->vo, "vaExportSurfaceHandle: composed layers not supported.\n");
+    if (CHECK_VA_STATUS(entry->vo, "vaExportSurfaceHandle()")) {
```

Comment:
I think this function is ripe for handling the errors with early exit.

What do you think of something a la https://github.com/jeeb/mpv/commit/bdee76e770c7e8a0f3b1b0d67ab3970c2f1159e6 ?

This loses a few levels of indentation and takes care of the else. Of course the cleanup goto is not perfect, but I think it's kept to a relatively simple level :) .